kong/plugins/bot-detection/handler.lua

```diff
@@ -0,0 +1,71 @@
+local BasePlugin = require "kong.plugins.base_plugin"
+local reg = require "rex_pcre"
+local responses = require "kong.tools.responses"
+local rules = require "kong.plugins.bot-detection.rules"
+local stringy = require "stringy"
+local bot_cache = require "kong.plugins.bot-detection.cache"
+
+local ipairs = ipairs
+
+local BotDetectionHandler = BasePlugin:extend()
+
+BotDetectionHandler.PRIORITY = 2500
+
+local ALLOW = 1
+local BLOCK = 2
+
+function BotDetectionHandler:new()
+  BotDetectionHandler.super.new(self, "bot-detection")
+end
+
+function BotDetectionHandler:access(conf)
+  BotDetectionHandler.super.access(self)
+
+  local user_agent = ngx.req.get_headers()["user-agent"]
```

Comment:
An option would be to wrap the request and response object in easy to use Kong objects, with some magic happening inside. If we did so, creating a custom plugin would be very easy, ie:

```lua
function BotDetectionHandler:access(conf, req, res)

req.querystring["hello"]
req.param["form_param_name"]
-- Sets a new param to the upstream request
req.param["another_param"] = "updated value"

-- Sets a new header by reading another header and appending "world"
res.header["x-custom"] = res.header["x-hello"].." world"

end
```

People with advanced requirements could still get the raw openresty request:

```
ngx.req
ngx.res
```